## Programming part of Homework 2 (Data Structures, Fall 2025)

## Name:陳彥岑
## Student ID Number:113820031

### Programming problem 1
Consider the problem to compute the product of two positive integers, $n$ and $m$, using only addition and subtraction.
This programming problem is to consider the recursive approach in above problem and asks you to complete the following tasks.
1. Implement the approach as a function named as `product_rec` using Python.
2. Please have an iterative version for the approach and write a function for this version with function name `product_ite`.
3. Compare these two function with the same input in terms of running time and write what you have observed.

**Note** that we will use an in-built python library **`timeit`** and the module function **`timeit.timeit()`** for measuring the running time.
For the programming problems, please do some experiments by yourself to observe the running time and the number of recursive calls. You need to report what you have observed from the experiments when you submit your homework.

In [3]:
def product_rec(n, m):
    if n == 1 :
        return m
    else:
        return n + product_rec(n - 1, m)
print("The product of 4 and 7 is", product_rec(4, 7))

The product of 4 and 7 is 16


In [5]:
def product_ite(n, m):
    result = 0
    for _ in range(n):
        result += m
    return result

print("The product of 4 and 7 is", product_ite(4, 7))

The product of 4 and 7 is 28


In [6]:
#---------------------------------------------------------timeit
import timeit as ti

#---------------------------------------------------------Method 1
print("Running time comparison (Timer Method 1):")
tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    product_rec(40, 700)
tend = ti.default_timer() #timer end計時器關閉
t_rec=tend - tstart#開始與結束的差值

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    product_ite(40, 700)
tend = ti.default_timer() #timer end計時器關閉
t_ite=tend - tstart#開始與結束的差值
print("Recursion 40*700=", product_rec(40, 700),"Iteration 40*700=", product_ite(40, 700))
print("Recursive approach:", t_rec)
print("iterative approach:", t_ite)

#---------------------------------------------------------Method 2
print("Running time comparison (Timer Method 2):")
t_rec = ti.timeit("product_rec(40, 700)", setup="from __main__ import product_rec", number = 10000)#timeit(函數名稱,執行次數)
t_ite = ti.timeit("product_ite(40, 700)", setup="from __main__ import product_ite", number = 10000)

print("Recursion 40*700=", product_rec(40, 700),"Iteration 40*700=", product_ite(40, 700))
print("Recursive approach:", t_rec)
print("Iterative approach:", t_ite)

Running time comparison (Timer Method 1):
Recursion 40*700= 1519 Iteration 40*700= 28000
Recursive approach: 0.02367759999469854
iterative approach: 0.01252399999066256
Running time comparison (Timer Method 2):
Recursion 40*700= 1519 Iteration 40*700= 28000
Recursive approach: 0.021500600007129833
Iterative approach: 0.011761200003093109


In [8]:
# n=80
#---------------------------------------------------------timeit
import timeit as ti

#---------------------------------------------------------Method 1
print("Running time comparison (Timer Method 1):")
tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    product_rec(80, 700)
tend = ti.default_timer() #timer end計時器關閉
t_rec=tend - tstart#開始與結束的差值

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    product_ite(80, 700)
tend = ti.default_timer() #timer end計時器關閉
t_ite=tend - tstart#開始與結束的差值
print("Recursion 80*700=", product_rec(80, 700),"Iteration 80*700=", product_ite(80, 700))
print("Recursive approach:", t_rec)
print("iterative approach:", t_ite)

#---------------------------------------------------------Method 2
print("Running time comparison (Timer Method 2):")
t_rec = ti.timeit("product_rec(80, 700)", setup="from __main__ import product_rec", number = 10000)#timeit(函數名稱,執行次數)
t_ite = ti.timeit("product_ite(80, 700)", setup="from __main__ import product_ite", number = 10000)

print("Recursion 80*700=", product_rec(80, 700),"Iteration 80*700=", product_ite(80, 700))
print("Recursive approach:", t_rec)
print("Iterative approach:", t_ite)

Running time comparison (Timer Method 1):
Recursion 80*700= 3939 Iteration 80*700= 56000
Recursive approach: 0.04499290001695044
iterative approach: 0.02225680000265129
Running time comparison (Timer Method 2):
Recursion 80*700= 3939 Iteration 80*700= 56000
Recursive approach: 0.042110800015507266
Iterative approach: 0.021765299985418096


In [9]:
# n = 120
#---------------------------------------------------------timeit
import timeit as ti

#---------------------------------------------------------Method 1
print("Running time comparison (Timer Method 1):")
tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    product_rec(120, 700)
tend = ti.default_timer() #timer end計時器關閉
t_rec=tend - tstart#開始與結束的差值

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    product_ite(120, 700)
tend = ti.default_timer() #timer end計時器關閉
t_ite=tend - tstart#開始與結束的差值
print("Recursion 120*700=", product_rec(120, 700),"Iteration 120*700=", product_ite(120, 700))
print("Recursive approach:", t_rec)
print("iterative approach:", t_ite)

#---------------------------------------------------------Method 2
print("Running time comparison (Timer Method 2):")
t_rec = ti.timeit("product_rec(120, 700)", setup="from __main__ import product_rec", number = 10000)#timeit(函數名稱,執行次數)
t_ite = ti.timeit("product_ite(120, 700)", setup="from __main__ import product_ite", number = 10000)

print("Recursion 120*700=", product_rec(120, 700),"Iteration 120*700=", product_ite(120, 700))
print("Recursive approach:", t_rec)
print("Iterative approach:", t_ite)

Running time comparison (Timer Method 1):
Recursion 120*700= 7959 Iteration 120*700= 84000
Recursive approach: 0.10187509999377653
iterative approach: 0.03178479999769479
Running time comparison (Timer Method 2):
Recursion 120*700= 7959 Iteration 120*700= 84000
Recursive approach: 0.09558699998888187
Iterative approach: 0.03208410000661388


In [10]:
# n = 200
#---------------------------------------------------------timeit
import timeit as ti

#---------------------------------------------------------Method 1
print("Running time comparison (Timer Method 1):")
tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    product_rec(200, 700)
tend = ti.default_timer() #timer end計時器關閉
t_rec=tend - tstart#開始與結束的差值

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    product_ite(200, 700)
tend = ti.default_timer() #timer end計時器關閉
t_ite=tend - tstart#開始與結束的差值
print("Recursion 200*700=", product_rec(200, 700),"Iteration 200*700=", product_ite(200, 700))
print("Recursive approach:", t_rec)
print("iterative approach:", t_ite)

#---------------------------------------------------------Method 2
print("Running time comparison (Timer Method 2):")
t_rec = ti.timeit("product_rec(200, 700)", setup="from __main__ import product_rec", number = 10000)#timeit(函數名稱,執行次數)
t_ite = ti.timeit("product_ite(200, 700)", setup="from __main__ import product_ite", number = 10000)

print("Recursion 200*700=", product_rec(200, 700),"Iteration 200*700=", product_ite(200, 700))
print("Recursive approach:", t_rec)
print("Iterative approach:", t_ite)

Running time comparison (Timer Method 1):
Recursion 200*700= 20799 Iteration 200*700= 140000
Recursive approach: 0.18229619998601265
iterative approach: 0.05492289998801425
Running time comparison (Timer Method 2):
Recursion 200*700= 20799 Iteration 200*700= 140000
Recursive approach: 0.17913720000069588
Iterative approach: 0.053587500005960464


In [11]:
# n = 2000
#---------------------------------------------------------timeit
import timeit as ti

#---------------------------------------------------------Method 1
print("Running time comparison (Timer Method 1):")
tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    product_rec(2000, 700)
tend = ti.default_timer() #timer end計時器關閉
t_rec=tend - tstart#開始與結束的差值

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    product_ite(2000, 700)
tend = ti.default_timer() #timer end計時器關閉
t_ite=tend - tstart#開始與結束的差值
print("Recursion 200*700=", product_rec(2000, 700),"Iteration 200*700=", product_ite(2000, 700))
print("Recursive approach:", t_rec)
print("iterative approach:", t_ite)

#---------------------------------------------------------Method 2
print("Running time comparison (Timer Method 2):")
t_rec = ti.timeit("product_rec(2000, 700)", setup="from __main__ import product_rec", number = 10000)#timeit(函數名稱,執行次數)
t_ite = ti.timeit("product_ite(2000, 700)", setup="from __main__ import product_ite", number = 10000)

print("Recursion 2000*700=", product_rec(2000, 700),"Iteration 2000*700=", product_ite(2000, 700))
print("Recursive approach:", t_rec)
print("Iterative approach:", t_ite)

Running time comparison (Timer Method 1):
Recursion 200*700= 2001699 Iteration 200*700= 1400000
Recursive approach: 2.7301742999989074
iterative approach: 0.798657400009688
Running time comparison (Timer Method 2):
Recursion 2000*700= 2001699 Iteration 2000*700= 1400000
Recursive approach: 2.689667800004827
Iterative approach: 0.7428969999891706


#### Please write down the observations you have after the comparisons:
##### Inputs (values of n) for five experiments
1. n = 40<br>
2. n = 80<br>
3. n = 120<br>
4. n = 200<br>
5. n = 2000<br>
##### Results of comparisoms
*******************************************************************************************************
**Method 1**

| time          |    time for exp1    |    time for exp2    |    time for exp3    |    time for exp4    |   time for exp5   | 
| -----------   |  ------------------ | ------------------- | ------------------- | ------------------- | ----------------- |
|Iteration      | 0.01252399999066256 | 0.02225680000265129 | 0.03178479999769479 | 0.05492289998801425 | 0.798657400009688 |
|Recursion      | 0.02367759999469854 | 0.04499290001695044 | 0.10187509999377653 | 0.18229619998601265 | 2.730174299998907 |

**Method 2**

| time         |    time for exp1     |    time for exp2     |    time for exp3    |    time for exp4    |   time for exp5   | 
| -----------  | -------------------- | -------------------- | ------------------- | ------------------- | ----------------- |
|Iteration     | 0.011761200003093109 | 0.021765299985418096 | 0.03208410000661388 | 0.05358750000596046 | 0.742896999989170 |
|Recursion     | 0.021500600007129833 | 0.042110800015507266 | 0.09558699998888187 | 0.17913720000069588 | 2.689667800004827 |

*******************************************************************************************************

##### Observations


### Programming problem 2
Consider the problem of ***Chebyshev polynomials of the first kind***. Please use Python to implement the pseudo-code you provide for solving the problem. You should provide all the functions with the mentioned names, `chebyshev_recursive`, `chebyshev_memoreccu`, `chebyshev_addoutrec`, and `chebyshev_iterative` respectively. Then, perform a comparison on these functions by measuring the execution time.
1. Please have the recursive function, `chebyshev_recursive` to list all the polynomials with the recursive definition.
2. Please provide the memoized recursive function, `chebyshev_memoreccu` to list all the polynomials.
3. Please use the way`LinearFibonacci}` we introduced in class to have a linear recursive function `chebyshev_addoutrec`.
4. Please provide an iterative (no recursive call) function, `chebyshev_iterative`, which lists all the polynomials iteratively.
5. In order to have the experiments for comparisons, please execute all the functions with the same input many times and then measure the practical running timme in average. Please write down your observations on these four versions.

**Note** that we will use an in-built python library **`timeit`** and the module function **`timeit.timeit()`** for measuring the running time.
For the programming problems, please do some experiments by yourself to observe the running time and the number of recursive calls. You need to report what you have observed from the experiments when you submit your homework.


In [5]:
def chebyshev_recursive(n):
    """Return coefficients of Chebyshev polynomial T_n(x) (first kind) recursively."""
    if n == 0:
        return [1]        # T0(x) = 1
    elif n == 1:
        return [1, 0]     # T1(x) = x

    Tn_1 = chebyshev_recursive(n - 1)
    Tn_2 = chebyshev_recursive(n - 2)

    Tn_1_shifted = [2 * c for c in Tn_1] + [0]


    len_diff = len(Tn_1_shifted) - len(Tn_2)
    if len_diff > 0:
        Tn_2 = [0] * len_diff + Tn_2
    elif len_diff < 0:
        Tn_1_shifted = [0] * (-len_diff) + Tn_1_shifted


    Tn = [a - b for a, b in zip(Tn_1_shifted, Tn_2)]
    return Tn  


if __name__ == "__main__":
    for n in range(11):
        print(f"The coefficients of T_{n}(x) in list = {chebyshev_recursive(n)}")


The coefficients of T_0(x) in list = [1]
The coefficients of T_1(x) in list = [1, 0]
The coefficients of T_2(x) in list = [2, 0, -1]
The coefficients of T_3(x) in list = [4, 0, -3, 0]
The coefficients of T_4(x) in list = [8, 0, -8, 0, 1]
The coefficients of T_5(x) in list = [16, 0, -20, 0, 5, 0]
The coefficients of T_6(x) in list = [32, 0, -48, 0, 18, 0, -1]
The coefficients of T_7(x) in list = [64, 0, -112, 0, 56, 0, -7, 0]
The coefficients of T_8(x) in list = [128, 0, -256, 0, 160, 0, -32, 0, 1]
The coefficients of T_9(x) in list = [256, 0, -576, 0, 432, 0, -120, 0, 9, 0]
The coefficients of T_10(x) in list = [512, 0, -1280, 0, 1120, 0, -400, 0, 50, 0, -1]


In [6]:
def chebyshev_memoreccu(n, memo):
    if memo[n] is not None:
        return memo[n]

    if n == 0:
        memo[0] = [1]        # T0(x) = 1
        return memo[0]
    if n == 1:
        memo[1] = [1, 0]     # T1(x) = x
        return memo[1]

    # 遞迴（使用快取）
    Tn_1 = chebyshev_memoreccu(n - 1, memo)  # T_{n-1}
    Tn_2 = chebyshev_memoreccu(n - 2, memo)  # T_{n-2}

    two_x_Tn1 = [c * 2 for c in Tn_1] + [0]

    L = max(len(two_x_Tn1), len(Tn_2))
    a = [0] * (L - len(two_x_Tn1)) + two_x_Tn1
    b = [0] * (L - len(Tn_2))     + Tn_2

    Tn = [ai - bi for ai, bi in zip(a, b)]
    memo[n] = Tn
    return Tn


if __name__ == "__main__":
    for n in range(11):
        m_table = [None] * (n + 1)    
        coeffs = chebyshev_memoreccu(n, m_table)
        print(f"The coefficients of T_{n}(x) in list = {coeffs}")


The coefficients of T_0(x) in list = [1]
The coefficients of T_1(x) in list = [1, 0]
The coefficients of T_2(x) in list = [2, 0, -1]
The coefficients of T_3(x) in list = [4, 0, -3, 0]
The coefficients of T_4(x) in list = [8, 0, -8, 0, 1]
The coefficients of T_5(x) in list = [16, 0, -20, 0, 5, 0]
The coefficients of T_6(x) in list = [32, 0, -48, 0, 18, 0, -1]
The coefficients of T_7(x) in list = [64, 0, -112, 0, 56, 0, -7, 0]
The coefficients of T_8(x) in list = [128, 0, -256, 0, 160, 0, -32, 0, 1]
The coefficients of T_9(x) in list = [256, 0, -576, 0, 432, 0, -120, 0, 9, 0]
The coefficients of T_10(x) in list = [512, 0, -1280, 0, 1120, 0, -400, 0, 50, 0, -1]


In [7]:
def chebyshev_addoutrec(n, Tn_1=None, Tn_2=None, k=1):
    if n == 0:
        return [1]        # T0(x) = 1
    elif n == 1:
        return [1, 0]     # T1(x) = x

    if Tn_1 is None and Tn_2 is None:
        Tn_2 = [1]        # T0
        Tn_1 = [1, 0]     # T1
        k = 1

    if k == n:
        return Tn_1

    two_x_Tn1 = [c * 2 for c in Tn_1] + [0]

    L = max(len(two_x_Tn1), len(Tn_2))
    a = [0] * (L - len(two_x_Tn1)) + two_x_Tn1
    b = [0] * (L - len(Tn_2)) + Tn_2

    T_next = [ai - bi for ai, bi in zip(a, b)]
    return chebyshev_addoutrec(n, T_next, Tn_1, k + 1)


if __name__ == "__main__":
    for n in range(11):
        print(f"The coefficients of T_{n}(x) in list = {chebyshev_addoutrec(n)}")


The coefficients of T_0(x) in list = [1]
The coefficients of T_1(x) in list = [1, 0]
The coefficients of T_2(x) in list = [2, 0, -1]
The coefficients of T_3(x) in list = [4, 0, -3, 0]
The coefficients of T_4(x) in list = [8, 0, -8, 0, 1]
The coefficients of T_5(x) in list = [16, 0, -20, 0, 5, 0]
The coefficients of T_6(x) in list = [32, 0, -48, 0, 18, 0, -1]
The coefficients of T_7(x) in list = [64, 0, -112, 0, 56, 0, -7, 0]
The coefficients of T_8(x) in list = [128, 0, -256, 0, 160, 0, -32, 0, 1]
The coefficients of T_9(x) in list = [256, 0, -576, 0, 432, 0, -120, 0, 9, 0]
The coefficients of T_10(x) in list = [512, 0, -1280, 0, 1120, 0, -400, 0, 50, 0, -1]


In [8]:
def chebyshev_iterative(n):
    
    if n == 0:
        return [1]          # T0(x) = 1
    elif n == 1:
        return [1, 0]       # T1(x) = x

    Tn_2 = [1]              # T0
    Tn_1 = [1, 0]           # T1

    for k in range(2, n + 1):
        two_x_Tn1 = [c * 2 for c in Tn_1] + [0]

        L = max(len(two_x_Tn1), len(Tn_2))
        a = [0] * (L - len(two_x_Tn1)) + two_x_Tn1
        b = [0] * (L - len(Tn_2)) + Tn_2

        Tn = [ai - bi for ai, bi in zip(a, b)]

        Tn_2, Tn_1 = Tn_1, Tn

    return Tn_1 


if __name__ == "__main__":
    for n in range(11):
        print(f"The coefficients of T_{n}(x) in list = {chebyshev_iterative(n)}")


The coefficients of T_0(x) in list = [1]
The coefficients of T_1(x) in list = [1, 0]
The coefficients of T_2(x) in list = [2, 0, -1]
The coefficients of T_3(x) in list = [4, 0, -3, 0]
The coefficients of T_4(x) in list = [8, 0, -8, 0, 1]
The coefficients of T_5(x) in list = [16, 0, -20, 0, 5, 0]
The coefficients of T_6(x) in list = [32, 0, -48, 0, 18, 0, -1]
The coefficients of T_7(x) in list = [64, 0, -112, 0, 56, 0, -7, 0]
The coefficients of T_8(x) in list = [128, 0, -256, 0, 160, 0, -32, 0, 1]
The coefficients of T_9(x) in list = [256, 0, -576, 0, 432, 0, -120, 0, 9, 0]
The coefficients of T_10(x) in list = [512, 0, -1280, 0, 1120, 0, -400, 0, 50, 0, -1]


In [9]:
#---------------------------------------------------------timeit
import timeit as ti

n = 10
print("Running time comparison (Timer Method 1):")

#----------------------Method 1:
tstart = ti.default_timer()
for i in range(1000):
    chebyshev_iterative(n)
tend = ti.default_timer()
t_ite = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    chebyshev_recursive(n)
tend = ti.default_timer()
t_rec = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    m_table = [None] * (n + 1)
    chebyshev_memoreccu(n, m_table)
tend = ti.default_timer()
t_mem = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    chebyshev_addoutrec(n)
tend = ti.default_timer()
t_add = tend - tstart

print(f"Deriving T_{{{n}}}(x)")
print(f"{'Iterative approach':>21} : {t_ite:<20.12f}")
print(f"{'Recursive approach':>21} : {t_rec:<20.12f}")
print(f"{'Memoized Recursion':>21} : {t_mem:<20.12f}")
print(f"{'Addoutrec approach':>21} : {t_add:<20.12f}")

#----------------------Method 2
print("\nRunning time comparison (Timer Method 2):")

def iterative_test():
    chebyshev_iterative(n)

def recursive_test():
    chebyshev_recursive(n)

def memoreccu_test():
    m_table = [None] * (n + 1)
    chebyshev_memoreccu(n, m_table)

def addoutrec_test():
    chebyshev_addoutrec(n)

t_ite = ti.timeit("iterative_test()", setup="from __main__ import iterative_test", number=1000)
t_rec = ti.timeit("recursive_test()", setup="from __main__ import recursive_test", number=1000)
t_mem = ti.timeit("memoreccu_test()", setup="from __main__ import memoreccu_test", number=1000)
t_add = ti.timeit("addoutrec_test()", setup="from __main__ import addoutrec_test", number=1000)

print(f"Deriving T_{{{n}}}(x)")
print(f"{'Iterative approach':>21} : {t_ite:<20.12f}")
print(f"{'Recursive approach':>21} : {t_rec:<20.12f}")
print(f"{'Memoized Recursion':>21} : {t_mem:<20.12f}")
print(f"{'Addoutrec approach':>21} : {t_add:<20.12f}")


Running time comparison (Timer Method 1):
Deriving T_{10}(x)
   Iterative approach : 0.011553699995      
   Recursive approach : 0.076990000001      
   Memoized Recursion : 0.012341599999      
   Addoutrec approach : 0.011608500005      

Running time comparison (Timer Method 2):
Deriving T_{10}(x)
   Iterative approach : 0.010676300008      
   Recursive approach : 0.079145299998      
   Memoized Recursion : 0.012669899996      
   Addoutrec approach : 0.011718499998      


In [13]:
#---------------------------------------------------------timeit
import timeit as ti

n = 15
print("Running time comparison (Timer Method 1):")

#----------------------Method 1:
tstart = ti.default_timer()
for i in range(1000):
    chebyshev_iterative(n)
tend = ti.default_timer()
t_ite = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    chebyshev_recursive(n)
tend = ti.default_timer()
t_rec = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    m_table = [None] * (n + 1)
    chebyshev_memoreccu(n, m_table)
tend = ti.default_timer()
t_mem = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    chebyshev_addoutrec(n)
tend = ti.default_timer()
t_add = tend - tstart

print(f"Deriving T_{{{n}}}(x)")
print(f"{'Iterative approach':>21} : {t_ite:<20.12f}")
print(f"{'Recursive approach':>21} : {t_rec:<20.12f}")
print(f"{'Memoized Recursion':>21} : {t_mem:<20.12f}")
print(f"{'Addoutrec approach':>21} : {t_add:<20.12f}")

#----------------------Method 2
print("\nRunning time comparison (Timer Method 2):")

def iterative_test():
    chebyshev_iterative(n)

def recursive_test():
    chebyshev_recursive(n)

def memoreccu_test():
    m_table = [None] * (n + 1)
    chebyshev_memoreccu(n, m_table)

def addoutrec_test():
    chebyshev_addoutrec(n)

t_ite = ti.timeit("iterative_test()", setup="from __main__ import iterative_test", number=1000)
t_rec = ti.timeit("recursive_test()", setup="from __main__ import recursive_test", number=1000)
t_mem = ti.timeit("memoreccu_test()", setup="from __main__ import memoreccu_test", number=1000)
t_add = ti.timeit("addoutrec_test()", setup="from __main__ import addoutrec_test", number=1000)

print(f"Deriving T_{{{n}}}(x)")
print(f"{'Iterative approach':>21} : {t_ite:<20.12f}")
print(f"{'Recursive approach':>21} : {t_rec:<20.12f}")
print(f"{'Memoized Recursion':>21} : {t_mem:<20.12f}")
print(f"{'Addoutrec approach':>21} : {t_add:<20.12f}")


Running time comparison (Timer Method 1):
Deriving T_{15}(x)
   Iterative approach : 0.026480800007      
   Recursive approach : 1.181851800007      
   Memoized Recursion : 0.031235900009      
   Addoutrec approach : 0.031737000012      

Running time comparison (Timer Method 2):
Deriving T_{15}(x)
   Iterative approach : 0.024641299999      
   Recursive approach : 1.121451499988      
   Memoized Recursion : 0.024666000012      
   Addoutrec approach : 0.025430999987      


In [10]:
#---------------------------------------------------------timeit
import timeit as ti

n = 20
print("Running time comparison (Timer Method 1):")

#----------------------Method 1:
tstart = ti.default_timer()
for i in range(1000):
    chebyshev_iterative(n)
tend = ti.default_timer()
t_ite = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    chebyshev_recursive(n)
tend = ti.default_timer()
t_rec = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    m_table = [None] * (n + 1)
    chebyshev_memoreccu(n, m_table)
tend = ti.default_timer()
t_mem = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    chebyshev_addoutrec(n)
tend = ti.default_timer()
t_add = tend - tstart

print(f"Deriving T_{{{n}}}(x)")
print(f"{'Iterative approach':>21} : {t_ite:<20.12f}")
print(f"{'Recursive approach':>21} : {t_rec:<20.12f}")
print(f"{'Memoized Recursion':>21} : {t_mem:<20.12f}")
print(f"{'Addoutrec approach':>21} : {t_add:<20.12f}")

#----------------------Method 2
print("\nRunning time comparison (Timer Method 2):")

def iterative_test():
    chebyshev_iterative(n)

def recursive_test():
    chebyshev_recursive(n)

def memoreccu_test():
    m_table = [None] * (n + 1)
    chebyshev_memoreccu(n, m_table)

def addoutrec_test():
    chebyshev_addoutrec(n)

t_ite = ti.timeit("iterative_test()", setup="from __main__ import iterative_test", number=1000)
t_rec = ti.timeit("recursive_test()", setup="from __main__ import recursive_test", number=1000)
t_mem = ti.timeit("memoreccu_test()", setup="from __main__ import memoreccu_test", number=1000)
t_add = ti.timeit("addoutrec_test()", setup="from __main__ import addoutrec_test", number=1000)

print(f"Deriving T_{{{n}}}(x)")
print(f"{'Iterative approach':>21} : {t_ite:<20.12f}")
print(f"{'Recursive approach':>21} : {t_rec:<20.12f}")
print(f"{'Memoized Recursion':>21} : {t_mem:<20.12f}")
print(f"{'Addoutrec approach':>21} : {t_add:<20.12f}")


Running time comparison (Timer Method 1):
Deriving T_{20}(x)
   Iterative approach : 0.029841999989      
   Recursive approach : 10.091054400022     
   Memoized Recursion : 0.033837199997      
   Addoutrec approach : 0.031341399997      

Running time comparison (Timer Method 2):
Deriving T_{20}(x)
   Iterative approach : 0.030617899989      
   Recursive approach : 14.020931100007     
   Memoized Recursion : 0.046907899989      
   Addoutrec approach : 0.046146500012      


In [11]:
#---------------------------------------------------------timeit
import timeit as ti

n = 25

print("Running time comparison (Timer Method 1):")

#----------------------Method 1:
tstart = ti.default_timer()
for i in range(1000):
    chebyshev_iterative(n)
tend = ti.default_timer()
t_ite = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    chebyshev_recursive(n)
tend = ti.default_timer()
t_rec = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    m_table = [None] * (n + 1)
    chebyshev_memoreccu(n, m_table)
tend = ti.default_timer()
t_mem = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    chebyshev_addoutrec(n)
tend = ti.default_timer()
t_add = tend - tstart

print(f"Deriving T_{{{n}}}(x)")
print(f"{'Iterative approach':>21} : {t_ite:<20.12f}")
print(f"{'Recursive approach':>21} : {t_rec:<20.12f}")
print(f"{'Memoized Recursion':>21} : {t_mem:<20.12f}")
print(f"{'Addoutrec approach':>21} : {t_add:<20.12f}")

#----------------------Method 2
print("\nRunning time comparison (Timer Method 2):")

def iterative_test():
    chebyshev_iterative(n)

def recursive_test():
    chebyshev_recursive(n)

def memoreccu_test():
    m_table = [None] * (n + 1)
    chebyshev_memoreccu(n, m_table)

def addoutrec_test():
    chebyshev_addoutrec(n)

t_ite = ti.timeit("iterative_test()", setup="from __main__ import iterative_test", number=1000)
t_rec = ti.timeit("recursive_test()", setup="from __main__ import recursive_test", number=1000)
t_mem = ti.timeit("memoreccu_test()", setup="from __main__ import memoreccu_test", number=1000)
t_add = ti.timeit("addoutrec_test()", setup="from __main__ import addoutrec_test", number=1000)

print(f"Deriving T_{{{n}}}(x)")
print(f"{'Iterative approach':>21} : {t_ite:<20.12f}")
print(f"{'Recursive approach':>21} : {t_rec:<20.12f}")
print(f"{'Memoized Recursion':>21} : {t_mem:<20.12f}")
print(f"{'Addoutrec approach':>21} : {t_add:<20.12f}")


Running time comparison (Timer Method 1):
Deriving T_{25}(x)
   Iterative approach : 0.041872999980      
   Recursive approach : 129.386125499994    
   Memoized Recursion : 0.074498599977      
   Addoutrec approach : 0.069451400021      

Running time comparison (Timer Method 2):
Deriving T_{25}(x)
   Iterative approach : 0.065336199972      
   Recursive approach : 119.411895700003    
   Memoized Recursion : 0.053034600016      
   Addoutrec approach : 0.050637299981      


In [12]:
#---------------------------------------------------------timeit
import timeit as ti

n = 27

print("Running time comparison (Timer Method 1):")

#----------------------Method 1:
tstart = ti.default_timer()
for i in range(1000):
    chebyshev_iterative(n)
tend = ti.default_timer()
t_ite = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    chebyshev_recursive(n)
tend = ti.default_timer()
t_rec = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    m_table = [None] * (n + 1)
    chebyshev_memoreccu(n, m_table)
tend = ti.default_timer()
t_mem = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    chebyshev_addoutrec(n)
tend = ti.default_timer()
t_add = tend - tstart

print(f"Deriving T_{{{n}}}(x)")
print(f"{'Iterative approach':>21} : {t_ite:<20.12f}")
print(f"{'Recursive approach':>21} : {t_rec:<20.12f}")
print(f"{'Memoized Recursion':>21} : {t_mem:<20.12f}")
print(f"{'Addoutrec approach':>21} : {t_add:<20.12f}")

#----------------------Method 2
print("\nRunning time comparison (Timer Method 2):")

def iterative_test():
    chebyshev_iterative(n)

def recursive_test():
    chebyshev_recursive(n)

def memoreccu_test():
    m_table = [None] * (n + 1)
    chebyshev_memoreccu(n, m_table)

def addoutrec_test():
    chebyshev_addoutrec(n)

t_ite = ti.timeit("iterative_test()", setup="from __main__ import iterative_test", number=1000)
t_rec = ti.timeit("recursive_test()", setup="from __main__ import recursive_test", number=1000)
t_mem = ti.timeit("memoreccu_test()", setup="from __main__ import memoreccu_test", number=1000)
t_add = ti.timeit("addoutrec_test()", setup="from __main__ import addoutrec_test", number=1000)

print(f"Deriving T_{{{n}}}(x)")
print(f"{'Iterative approach':>21} : {t_ite:<20.12f}")
print(f"{'Recursive approach':>21} : {t_rec:<20.12f}")
print(f"{'Memoized Recursion':>21} : {t_mem:<20.12f}")
print(f"{'Addoutrec approach':>21} : {t_add:<20.12f}")


Running time comparison (Timer Method 1):
Deriving T_{27}(x)
   Iterative approach : 0.052774299984      
   Recursive approach : 451.576018499996    
   Memoized Recursion : 0.082234300004      
   Addoutrec approach : 0.090481399995      

Running time comparison (Timer Method 2):
Deriving T_{27}(x)
   Iterative approach : 0.092036299990      
   Recursive approach : 458.066367099993    
   Memoized Recursion : 0.085580699990      
   Addoutrec approach : 0.080823400000      


In [15]:
#---------------------------------------------------------timeit
import timeit as ti

n = 30

print("Running time comparison (Timer Method 1):")

#----------------------Method 1:
tstart = ti.default_timer()
for i in range(1000):
    chebyshev_iterative(n)
tend = ti.default_timer()
t_ite = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    chebyshev_recursive(n)
tend = ti.default_timer()
t_rec = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    m_table = [None] * (n + 1)
    chebyshev_memoreccu(n, m_table)
tend = ti.default_timer()
t_mem = tend - tstart

tstart = ti.default_timer()
for i in range(1000):
    chebyshev_addoutrec(n)
tend = ti.default_timer()
t_add = tend - tstart

print(f"Deriving T_{{{n}}}(x)")
print(f"{'Iterative approach':>21} : {t_ite:<20.12f}")
print(f"{'Recursive approach':>21} : {t_rec:<20.12f}")
print(f"{'Memoized Recursion':>21} : {t_mem:<20.12f}")
print(f"{'Addoutrec approach':>21} : {t_add:<20.12f}")

#----------------------Method 2
print("\nRunning time comparison (Timer Method 2):")

def iterative_test():
    chebyshev_iterative(n)

def recursive_test():
    chebyshev_recursive(n)

def memoreccu_test():
    m_table = [None] * (n + 1)
    chebyshev_memoreccu(n, m_table)

def addoutrec_test():
    chebyshev_addoutrec(n)

t_ite = ti.timeit("iterative_test()", setup="from __main__ import iterative_test", number=1000)
t_rec = ti.timeit("recursive_test()", setup="from __main__ import recursive_test", number=1000)
t_mem = ti.timeit("memoreccu_test()", setup="from __main__ import memoreccu_test", number=1000)
t_add = ti.timeit("addoutrec_test()", setup="from __main__ import addoutrec_test", number=1000)

print(f"Deriving T_{{{n}}}(x)")
print(f"{'Iterative approach':>21} : {t_ite:<20.12f}")
print(f"{'Recursive approach':>21} : {t_rec:<20.12f}")
print(f"{'Memoized Recursion':>21} : {t_mem:<20.12f}")
print(f"{'Addoutrec approach':>21} : {t_add:<20.12f}")


: 

#### Please write down the observations you have after the comparisons:
##### Inputs for five experiments
1. n = 10<br>
2. n = 15<br>
3. n = 20<br>
4. n = 25<br>
5. n = 27<br>
##### Results of comparisoms
*******************************************************************************************************
**Method 1**

| time          | time for exp1  | time for exp2  | time for exp3  |  time for exp4   |  time for exp5   | 
| -----------   | -------------- | -------------- | -------------- | ---------------- | ---------------- |
|Iteration      | 0.012064599985 | 0.026480800007 | 0.029841999989 | 0.041872999980   | 0.052774299984   |
|Recursion      | 0.075527900015 | 1.181851800007 | 10.09105440002 | 129.386125499994 | 451.576018499996 |
|Memoized       | 0.012863499986 | 0.031235900009 | 0.033837199997 | 0.074498599977   | 0.082234300004   |
|Addoutrec      | 0.010621499998 | 0.031737000012 | 0.031341399997 | 0.069451400021   | 0.090481399995   |

**Method 2**

| time          | time for exp1  | time for exp2  | time for exp3  |  time for exp4   |  time for exp5   | 
| -----------   | -------------- | -------------- | -------------- | ---------------- | ---------------- |
|Iteration      | 0.010944000009 | 0.024641299999 | 0.030617899989 | 0.065336199972   | 0.092036299990   |
|Recursion      | 0.081796000013 | 1.121451499988 | 14.02093110000 | 119.411895700003 | 458.066367099993 |
|Memoized       | 0.012787800020 | 0.024666000012 | 0.046907899989 | 0.053034600016   | 0.085580699990   |
|Addoutrec      | 0.011928599997 | 0.025430999987 | 0.046146500012 | 0.050637299981   | 0.080823400000   |

*******************************************************************************************************

##### Observations


1. **整體趨勢一致**  

   不論是使用 **Method 1** 或 **Method 2** 進行計時，四種演算法的執行時間變化趨勢相同：  
   > 「遞迴 (Recursion) ≫ 其他三種方法」  
   這顯示兩種計時方式的結果具有一致性與可靠性。

*******************************************************************************************************

2. **遞迴 (Recursion) 的效率最低**  
   
   隨著輸入規模增加（由 exp1 至 exp5），遞迴法的時間呈現 **指數級成長**，在 exp5 甚至超過 100 秒。  
   這是因為遞迴法在沒有任何記憶化機制時，會重複計算相同子問題，導致計算量暴增。

*******************************************************************************************************

3. **迭代 (Iteration)**、**記憶化遞迴 (Memoized)** 與 **加法式遞迴 (Addoutrec)** 的表現相近 

   這三種方法的執行時間皆落在相同數量級（約 0.01 至 0.09 秒），  
   顯示這些方法在時間效率上相當，且執行效率遠優於單純的遞迴。

*******************************************************************************************************

4. **Method 1 與 Method 2 的數值略有差異**  

   個別實驗的絕對數值並不完全一致，但差距不大，  
   推測可能來自系統背景負載、計時方式（例如 `timeit` 或 `default_timer`）的誤差，  
   整體趨勢仍具一致性，因此可視為兩者在準確度上相符。